penty/pentypes/builtins.py

```diff
@@ -529,13 +532,205 @@ def dict_instanciate(ty):
 ##
 #
 
+def check_self_ty(expected_ty):
+    def decorator(fun):
+        def wrapper(self_ty, *args, **kwargs):
+            if not issubclass(self_ty, expected_ty):
+                raise TypeError
+            return fun(self_ty, *args, **kwargs)
+        return wrapper
+    return decorator
+
+
+@check_self_ty(list)
+def list_add(self_ty, other_ty):
+    if not issubclass(other_ty, list):
+        raise TypeError
+    return _List[self_ty.__args__[0].union(other_ty.__args__[0])]
+
+@check_self_ty(list)
+def list_bool(self_ty):
+    if self_ty.__args__[0] == set():
+        return _Cst[False]
+    return bool
+
+@check_self_ty(list)
 def list_append(self_ty, value_ty):
     self_ty.__args__[0].add(_astype(value_ty))
     return _Cst[None]
 
+@check_self_ty(list)
+def list_contains(self_ty, other_ty):
+    # We cannot compare self_ty.__args__[0] with other_ty.__args__[0] for non
+    # empty intersection since values can evaluate equal while having different
+    # types (1 == 1.0)
+    return bool
+
+@check_self_ty(list)
+def list_delitem(self_ty, idx_ty):
+    if not issubclass(idx_ty, int):
```

Comment:
interesting one. we should also check if ``idx_ty`` as an ``__index__`` method. And you should use ``_astype`` on ``idx_ty``